### PR TITLE
feat(telemetry): send txpool.import telemetry msg

### LINKED ISSUE #1836: Implement `txpool.import` Telemetry Interface

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes

- From this commit, a telemetry message of type `txpool.import` will be
sent when a new transaction is imported in the transaction pool.

- Added tests for `txpool.import` and improved TestHandler_SendMulti test.

## Tests

- Setup a local telemetry server using substrate-telemetry.

Use the branch `print-payload-type` from my fork `https://github.com/kishansagathiya/substrate-telemetry` (It print the message received, type of message and error. From what I have played around upstream doesn't do that.)

Run telementry backend (`telemetry_core` and `telemetry_shard`).

-  Run gossamer by setting telemetry  url to the local server. 

Run `./bin/gossamer --chain polkadot --name 'Kishans Test Node' --telemetry-url 'ws://localhost:8001/submit 0'`

Follow the logs of telemetry shard. Wait for few minutes. Eventually, you should see logs like below
```
telemetry message received: {"best":"0x8f9e20e28e1a5b5fb1b9011e1b60a7fdf42fb610c0f4ebf6e5d9739ced7648ee","height":1645,"msg":"block.import","origin":"NetworkInitialSync","ts":"2021-11-09T18:34:05.96264119+05:30"}
message type Payload::BlockImport

```

## Issues

- Closes #1836

## Primary Reviewer

- @noot

### CODE REVIEW

File: dot/network/service.go [hunk 1]
```diff
@@ -333,12 +333,12 @@ main:
 			o := s.host.bwc.GetBandwidthTotals()
 			err := telemetry.GetInstance().SendMessage(telemetry.NewBandwidthTM(o.RateIn, o.RateOut, s.host.peerCount()))
 			if err != nil {
-				logger.Debug("problem sending system.interval telemetry message", "error", err)
+				logger.Warn("problem sending system.interval telemetry message", "error", err)
```
Comment (noot) [inline dot/network/service.go:336]: can we leave these as debug? I saw these getting spammed before sometimes lol. same for the others

**Review by noot — APPROVED** [review]
looks good! pls just change logs back to debug

File: dot/network/service.go (hunk 1 above)
Comment (kishansagathiya) [reply dot/network/service.go]: addressed

File: dot/state/block_finalisation.go [hunk 2]
```diff
@@ -191,7 +191,7 @@ func (bs *BlockState) SetFinalisedHash(hash common.Hash, round, setID uint64) er
 		),
 	)
 	if err != nil {
-		return fmt.Errorf("could not send 'notify.finalized' telemetry message, error: %s", err)
+		logger.Debug("could not send 'notify.finalized' telemetry message, error: %s", err)
```
Comment (qdm12) [inline dot/state/block_finalisation.go:194]: Merge development in your branch and use `logger.Debugf` instead, or that will break the dev branch.

nit also maybe a bit shorter:

```suggestion
		logger.Debugf("could not send 'notify.finalized' telemetry message: %s", err)
```

File: dot/state/transaction.go [hunk 3]
```diff
@@ -68,7 +70,16 @@ func (s *TransactionState) RemoveExtrinsicFromPool(ext types.Extrinsic) {
 // AddToPool adds a transaction to the pool
 func (s *TransactionState) AddToPool(vt *transaction.ValidTransaction) common.Hash {
 	s.notifyStatus(vt.Extrinsic, transaction.Future)
-	return s.pool.Insert(vt)
+
+	hash := s.pool.Insert(vt)
+
+	if err := telemetry.GetInstance().SendMessage(
+		telemetry.NewTxpoolImportTM(uint(s.queue.Len()), uint(s.pool.Len())),
+	); err != nil {
```
Comment (qdm12) [inline dot/state/transaction.go:78]: nit please split this long line

```suggestion
	message := telemetry.NewTxpoolImportTM(uint(s.queue.Len()), uint(s.pool.Len()))
	err := telemetry.GetInstance().SendMessage(message)
	if err != nil {
```

File: dot/state/transaction.go [hunk 4]
```diff
@@ -68,7 +70,16 @@ func (s *TransactionState) RemoveExtrinsicFromPool(ext types.Extrinsic) {
...
+		logger.Debug("problem sending txpool.import telemetry message", "error", err)
```
Comment (qdm12) [inline dot/state/transaction.go:79]: Merge the dev branch and change to

```suggestion
		logger.Debugf("problem sending txpool.import telemetry message: %s", err)
```

nit also should we change that to a warning instead? :thinking:

File: dot/telemetry/telemetry_test.go [hunk 5]
```diff
@@ -42,75 +42,66 @@ func TestMain(m *testing.M) {
...
-		wg.Done()
-	}()
+	messages := []Message{
+		NewBandwidthTM(2, 3, 1),
+		NewTxpoolImportTM(1, 2),
+
+		func(genesisHash common.Hash) Message {
+			return NewSystemConnectedTM(false, "chain", &genesisHash,
+				"systemName", "nodeName", "netID", "startTime", "0.1")
+		}(common.MustHexToHash("0x91b171bb158e2d3848fa23a9f1c25182fb8e20313b2c1eb49219da7a70ce90c3")),
```
Comment (qdm12) [inline dot/telemetry/telemetry_test.go:62]: Why the wrapping function? That's quite complex, maybe put pre-defining arguments above `messages`.

File: dot/telemetry/telemetry_test.go [hunk 6]
```diff
@@ -42,75 +42,66 @@ func TestMain(m *testing.M) {
...
 	var actual [][]byte
 	for data := range resultCh {
 		actual = append(actual, data)
-		if len(actual) == 6 {
+		if len(actual) == len(expected) {
 			break
 		}
 	}
 
+	sort.Slice(expected, func(i, j int) bool {
```
Comment (qdm12) [inline dot/telemetry/telemetry_test.go:101]: I believe you can just use `assert.ElementsMatch(t, expected, actual)` in this case, if order doesn't matter.

File: dot/telemetry/txpool_import.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package telemetry
+
+// txpoolImportTM holds `txpool.import` telemetry message, which is supposed to be
+// sent when a new transaction gets imported in the transaction pool.
+type txpoolImportTM struct {
+	Ready  uint `json:"ready"`
+	Future uint `json:"future"`
+}
```
Comment (qdm12) [inline dot/telemetry/txpool_import.go:24]: Does it hold "`txpool.import` telemetry message"? I can only see Ready and Future uints :thinking:

File: dot/telemetry/txpool_import.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+// NewTxpoolImportTM gets a new txpoolImportTM struct
```
Comment (qdm12) [inline dot/telemetry/txpool_import.go:26]: nit get means to me get from a pool/somewhere vs create

```suggestion
// NewTxpoolImportTM creates a new txpoolImportTM struct
```

File: dot/state/block_finalisation.go (hunk 2 above)
Comment (kishansagathiya) [reply dot/state/block_finalisation.go]: done

File: dot/state/transaction.go (hunk 4 above)
Comment (kishansagathiya) [reply dot/state/transaction.go]: done

File: dot/telemetry/txpool_import.go (hunk 8 above)
Comment (kishansagathiya) [reply dot/telemetry/txpool_import.go]: done

File: dot/telemetry/txpool_import.go (hunk 7 above)
Comment (kishansagathiya) [reply dot/telemetry/txpool_import.go:24]: It is as per what telemetry messages substrate is sending,
```
			telemetry!(
				telemetry;
				SUBSTRATE_INFO;
				"txpool.import";
				"ready" => status.ready,
				"future" => status.future,
			);
```

**Review by timwu20 — APPROVED** [review]

**Pull request merged**